### ISSUE TITLE
Will the reasoning speed increase significantly after switching to onnx? Can it be done to the millimeter level?

### ISSUE THREAD

**Issue by WGS-note** [DakeQQ/F5-TTS-ONNX#13]
Will the reasoning speed increase significantly after switching to onnx? Can it be done to the millimeter level?

**Comment by DakeQQ** [comment]
F5-TTS-ONNX may not meet your expectations; however, for Windows users with an AMD GPU seeking a straightforward way to run F5-TTS, it can be a suitable option. Notably, even RTX 4090 + TensorRT, achieving millisecond performance levels is unrealistic.

**Comment by WGS-note** [comment]
Thank you very much for your reply, which open source TTS are capable of real time synthesis?

Thanks again for the reply, very much looking forward to it.

**Comment by DakeQQ** [comment]
These two TTS repositories have the potential for achieving real-time voice cloning with free commercial use, even on common hardware, including mobile devices.
1. [FireRedTTS](https://github.com/FireRedTeam/FireRedTTS)
2. [GPT-SoVITS](https://github.com/RVC-Boss/GPT-SoVITS)

**Comment by WGS-note** [comment]
Thank you very much for your reply! 

Wish you all the best!

**Comment by WGS-note** [comment]
Excuse me, I tried FireRedTTS.

For the same text：“你好，我这边是XXXX的工作人员，请问你的号码是八八九八八九七三八八九吗？”

With the same GPU, F5-TT takes only 3s while FireRedTTS takes 10s, why?

**Comment by DakeQQ** [comment]
I’m not sure what issue you’re facing, but my CPU inference is 3 times faster compared to F5-TTS. The NFE steps required in FireRedTTS are significantly fewer than those in F5-TTS to achieve the same quality. Additionally, the hidden size in the Transformer blocks is much smaller than in F5-TTS. That's why I believe it has the potential for real-time voice cloning.

**Comment by WGS-note** [comment]
In this code：

[https://github.com/FireRedTeam/FireRedTTS/blob/3c1fb92c25919996a6d16af3d740dc1d4181a8c9/fireredtts/fireredtts.py#L165](url)

My time additions are as follows:

```python
    def synthesize(self, prompt_wav, text, lang="auto"):
        """_summary_

        Args:
            prompt_wav (_type_): _description_
            text (_type_): _description_
            lang (str, optional): _description_. Defaults to "auto".
        """
        substrings = text_split(text=text)
        print("[DEBUG] len substrings: ", len(substrings))

        import time
        out_wavs = []
        for sub in substrings:

            s2s2 = time.time()
            chunk = self.synthesize_base(prompt_wav=prompt_wav, text=sub, lang=lang)
            print("[DEBUG] s2s2 use time: ", time.time() - s2s2)

            out_wavs.append(chunk.detach().cpu())
        out_wav = torch.concat(out_wavs, axis=-1)
        return out_wav
```


Thanks again for your reply, much appreciated!

**Comment by DakeQQ** [comment]
I ran it again and still achieved a result that is 3x faster on the CPU. I believe your test results might be influenced by the pre-processing and post-processing code. There’s no doubt that the transformer blocks are faster than F5-TTS.

You can separate and simplify the pre-processing and post-processing code in FireRedTTS to achieve a clean and fast voice cloning experience.

**Comment by WGS-note** [comment]
Really appreciate your reply

I managed to switch to onnx, but the reasoning time isn't fast either... shouldn't I switch to onnx when the reasoning time gets faster?

The console output is as follows:

```
Building prefix dict from the default dictionary ...
Loading model from cache /home/wangguisen/tmp/jieba.cache
Loading model cost 1.223 seconds.
Prefix dict has been built successfully.


Run F5-TTS by ONNX Runtime.
NFE_STEP: 0
NFE_STEP: 1
NFE_STEP: 2
NFE_STEP: 3
NFE_STEP: 4
NFE_STEP: 5
NFE_STEP: 6
NFE_STEP: 7
NFE_STEP: 8
NFE_STEP: 9
NFE_STEP: 10
NFE_STEP: 11
NFE_STEP: 12
NFE_STEP: 13
NFE_STEP: 14
NFE_STEP: 15
NFE_STEP: 16
NFE_STEP: 17
NFE_STEP: 18
NFE_STEP: 19
NFE_STEP: 20
NFE_STEP: 21
NFE_STEP: 22
NFE_STEP: 23
NFE_STEP: 24
NFE_STEP: 25
NFE_STEP: 26
NFE_STEP: 27
NFE_STEP: 28
NFE_STEP: 29
NFE_STEP: 30
NFE_STEP: 31

Audio generation is complete.

ONNXRuntime Time Cost in Seconds:
251.953
```

**Comment by DakeQQ** [comment]
It’s evident that GPU acceleration hasn’t been enabled for ONNX Runtime on your computer. To enable GPU support, follow the instructions on the official ONNX [website](https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html) and quantize the model to Float16 format (or [download URL](https://huggingface.co/huggingfacess/F5-TTS-ONNX)). For Windows users, installing `onnxruntime-directml` and setting the provider to `DmlExecutionProvider` offers a straightforward way to utilize the GPU. While it’s a known issue that running F5 with ONNX Runtime is slower than PyTorch despite extensive optimizations, ONNX with DirectML provides the advantage of easy GPU access on Windows, whereas other methods often require setting up multiple environments.

You can use TensorRT to convert the exported ONNX model into a specialized format for faster performance on NVIDIA devices. Alternatively, you can use OpenVINO to optimize the exported ONNX model as well.

In summary, this F5-ONNX version may not directly improve speed on the ONNX Runtime framework itself, but it does on others.

**Comment by WGS-note** [comment]
thank you for your reply.
